### ISSUE TITLE
KG2.8.3c rollout 

### ISSUE THREAD

**Issue by sundareswarpullela** [RTXteam/RTX#2056]
_NOTE: To create a new issue based on this template, simply go to: https://github.com/RTXteam/RTX/issues/new?template=kg2rollout.md_

**THE BRANCH FOR THIS ROLLOUT IS: `KG2.8.3c`**
**THE ARAX-DATABASES.RTX.AI DIRECTORY FOR THIS ROLLOUT IS: `/home/rtxconfig/KG2.8.3`**

#### 1. Build and load KG2c:

- [x] merge `master` into the branch being used for this KG2 version
- [x] update the four hardcoded biolink version numbers in the branch (as needed):
  - [x] in `code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml` ([github](https://github.com/RTXteam/RTX/tree/master/code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml#L18); [local](../code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml))
  - [x] in `code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml` ([github](https://github.com/RTXteam/RTX/tree/master/code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml#L18); [local](../code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml))
  - [x] in `code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml` ([github](https://github.com/RTXteam/RTX/tree/master/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml#L17); [local](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml))
  - [x] in `code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml` ([github](https://github.com/RTXteam/RTX/tree/master/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml#L17); [local](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml))
- [x] build a new KG2c on `buildkg2c.rtx.ai` from the branch (how-to is [here](https://github.com/RTXteam/RTX/tree/master/code/kg2c#build-kg2canonicalized))
  - [x] before starting the build:
    - [x] make sure there is enough disk space available on `arax-databases.rtx.ai` (need at least 100G, ideally >120G). delete old KG2 database directories as needed (warn the team on Slack in advance).
    - [x] make sure to choose to build a new synonymizer in `kg2c_config.json`, as described in the how-to
  - [x] after the build is done, verify it looks ok:
    - [x] `node_synonymizer.sqlite` should be around 8-15 GB
    - [x] make sure `node_synonymizer.sqlite`'s last modified date is today (or whatever day the build was run)
    - [x] make sure `kg2c_lite.json.gz`'s last modified date is today (or whatever day the build was run)
    - [x] the entire build runtime (synonymizer + KG2c) shouldn't have been more than 24 hours
    - [x] the synonymizer and KG2c artifacts should have been auto-uploaded into the proper directory on `arax-databases.rtx.ai` (`/home/rtxconfig/KG2.X.Y`)
- [x] load the new KG2c into neo4j at http://kg2-X-Yc.rtx.ai:7474/browser/ (how to is [here](https://github.com/RTXteam/RTX/tree/master/code/kg2c#host-kg2canonicalized-in-neo4j))
  - [x] verify the correct KG2 version was uploaded by running this query: `match (n {id:"RTX:KG2c"}) return n`
- [x] update `RTX/code/config_dbs.json` in the branch:
  - [x] update the synonymizer version number/path
  - [x] update the fda_approved_drugs version number/path
  - [x] update the autocomplete version number/path
  - [x] update the meta_kg version number/path
  - [x] update the kg2c sqlite version number/path
  - [x] update the KG2pre and KG2c Neo4j endpoints
- [ ] upload the new `kg2c_lite_2.X.Y.json.gz` file to the [translator-lfs-artifacts](https://github.com/ncats/translator-lfs-artifacts/tree/main/files) repo
- [ ] upload the new `kg2_nodes_not_in_sri_nn.tsv` file to the [translator-lfs-artifacts](https://github.com/ncats/translator-lfs-artifacts/tree/main/files) repo
- [x] load the new KG2c into Plover (how-to is [here](https://github.com/RTXteam/PloverDB/wiki/Deployment-how-tos#to-build-plover-from-a-new-kg2-version))
- [x] update `config_dbs.json` in the branch for this KG2 version in the RTX repo to point to the new Plover **for the 'dev' maturity level**

#### 2. Rebuild downstream databases:

The following databases should be rebuilt and copies of them should be put in `/home/rtxconfig/KG2.X.Y` on `arax-databases.rtx.ai`. Please use this kind of naming format: `mydatabase_v1.0_KG2.X.Y.sqlite`.

- [x] NGD database (how-to is [here](https://github.com/RTXteam/RTX/blob/master/code/ARAX/ARAXQuery/Overlay/ngd/README.md))
- [x] refreshed DTD @chunyuma
- [x] DTD model @chunyuma _(may be skipped - depends on the changes in this KG2 version)
- [x] DTD database @chunyuma _(may be skipped - depends on the changes in this KG2 version)
- [x] XDTD database @chunyuma

**NOTE**: As databases are rebuilt, `RTX/code/config_dbs.json` will need to be updated to point to their new paths! Push these changes to the branch for this KG2 version, unless the rollout of this KG2 version has already occurred, in which case you should push to `master` (but first follow the steps described [here](https://github.com/RTXteam/RTX/wiki/Config,-databases,-and-SFTP#config_dbsjson)). 

#### 3. Update the ARAX codebase:

All code changes should **go in the branch for this KG2 version**!

- [x] regenerate the KG2c test triples file in the branch for this KG2 version @acevedol
  - [x] ensure the new KG2c Neo4j is currently running
  - [x] check out the branch and pull to get the latest changes (this is important for ensuring the correct KG2c Neo4j is used)
  - [x] run [create_json_of_kp_predicate_triples.py](https://github.com/RTXteam/RTX/blob/master/code/ARAX/KnowledgeSources/create_json_of_kp_predicate_triples.py)
  - [x] push the regenerated file to `RTX/code/ARAX/KnowledgeSources/RTX_KG2c_test_triples.json`
- [x] update Expand code as needed
- [x] update any other modules as needed
- [x] test everything together:
  - [x] check out the branch and pull to get the latest changes
  - [x] locally set `force_local = True` in `ARAX_expander.py` (to avoid using the old KG2 API)
  - [x] then run the entire ARAX pytest suite (i.e., `pytest -v`)
  - [x] address any failing tests
- [x] update the KG2 and ARAX version numbers in the appropriate places (in the branch for this KG2 version)
  - [x] Bump version on line 12 in `RTX/code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml` ([github](https://github.com/RTXteam/RTX/blob/master/code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml#L12); [local](../code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml)); the major and minor release numbers are kept synchronous with the TRAPI version; just bump the patch release version (least significant digit)
  - [x] Bump version on line 12 in `RTX/code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml` ([github](https://github.com/RTXteam/RTX/blob/master/code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml#L12); [local](../code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml)); the first three digits are kept synchronous with the KG2 release version
  - [x] Bump version on line 4 in `RTX/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml` ([github](https://github.com/RTXteam/RTX/blob/master/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml#L4); [local](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml)); same as for the ARAX `openapi.yaml` file
  - [x] Bump version on line 4 in `RTX/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml` ([github](https://github.com/RTXteam/RTX/blob/master/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml#L4); [local](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml)); same as for the KG2 `openapi.yaml` file
  

#### 4. Pre-upload databases:

Before rolling out, we need to pre-upload the new databases (referenced in `config_dbs.json`) to `arax.ncats.io` and the ITRB SFTP server. These steps can be done well in advance of the rollout; it doesn't hurt anything to do them early.

- [x] make sure `arax.ncats.io` has at least 100G of disk space free; delete old KG2 databases to free up space as needed (warn the team on slack first)
- [x] copy the new databases from `arax-databases.rtx.ai` to `arax.ncats.io:/data/orangeboard/databases/KG2.X.Y`; example for KG2.8.0:
  - [x] `ssh myuser@arax.ncats.io`
  - [x] `cd /data/orangeboard/databases/`
  - [x] `mkdir -m 777 KG2.8.0`
  - [x] `scp rtxconfig@arax-databases.rtx.ai:/home/rtxconfig/KG2.8.0/*2.8.0* KG2.8.0/`
- [x] upload the new databases and their md5 checksums to ITRB's SFTP server using the steps detailed [here](https://github.com/RTXteam/RTX/wiki/Config,-databases,-and-SFTP#steps-for-all-databases-at-once)

#### 5. Do the rollout:

- [x] merge `master` into the branch for this KG2 version
- [x] merge the branch into `master`
- [x] roll `master` out to the various `arax.ncats.io` endpoints
- [x] run the database manager
- [ ] run the pytest suite on the various endpoints
- [x] verify the new KG2 version is actually being used
  - [x] run this JSON query in the ARAX/KG2 UIs: `{"nodes": {"n00": {"ids": ["RTX:KG2c"]}}, "edges": {}}`
  - [x] look up `RTX:KG2` in the Synonyms tab in the UI
- [x] update our CI/CD testing instance with the new databases:
  - [x] `ssh ubuntu@cicd.rtx.ai`
  - [x] `cd RTX`
  - [x] `git pull origin master`
  - [x] `sudo bash`
  - [x] `mkdir -m 777 /mnt/data/orangeboard/databases/KG2.X.Y`
  - [x] `exit`
  - [x] `python3 code/ARAX/ARAXQuery/ARAX_database_manager.py --mnt --skip-if-exists --remove_unused`

#### 6. Final items/clean up:

- [x] turn off the old KG2c version's neo4j instance
- [x] turn off the old KG2c version's plover instance
- [x] turn off the new KG2pre version's neo4j instance
- [x] upgrade the ITRB Plover endpoint (https://kg2cploverdb.ci.transltr.io) to this KG2 version and make the KG2 API start using it (instead of our self-hosted endpoint): 
    - [x] update `kg_config.json` in the `main` branch of the Plover repo to point to the new `kg2c_lite_2.X.Y.json.gz` file (push this change)
    - [x] wait about 45 minutes for the endpoint to rebuild and then run Plover tests to verify it's working
    - [x] run the ARAX pytest suite with the NCATS endpoint plugged in (locally change the URL in `config_dbs.json` and set `force_local = True` in Expand)
    - [x] if all tests pass, update `config_dbs.json` in `master` to point to the ITRB Plover endpoints (all maturity levels)
    - [x] roll `master` out to the various endpoints on arax.ncats.io
    - [x] turn off our plover endpoint and verify once more that ARAX is still working ok

**Comment by amykglen** [comment]
@sundareswarpullela - I added your RSA public key to `kg2cplover.rtx.ai`, `kg2cplover2.rtx.ai`, and `ngdbuild.rtx.ai`. 

I think it would work best if you do the Plover build tomorrow (Thursday 6/15), because later today the KG2 ITRB Test will start using the Plover ITRB Test instance again (instead of our own self-hosted Plover instance), which means you will be safe to load the new Plover onto `kg2cplover.rtx.ai` at that point.

**Comment by amykglen** [comment]
@sundareswarpullela - note that for the two steps that involve uploading files to the NCATS Git LFS repo, we will probably need to have Mark Williams add you to the CODEOWNERS for that repo: https://github.com/ncats/translator-lfs-artifacts/blob/main/CODEOWNERS (otherwise you won't be able to push to `main`). @saramsey might be able to help facilitate this?

**Comment by edeutsch** [comment]
Maybe we can try to roll this out at hackathon on 6/21?

**Comment by amykglen** [comment]
@sundareswarpullela - I just asked Mark in Slack to add you to the Git LFS CODEOWNERS (https://ncatstranslator.slack.com/archives/C023J4P0Y1M/p1686766927551099)

**Comment by saramsey** [comment]
Steve working on this issue in order to free up @sundareswarpullela to work on #2022

**Comment by saramsey** [comment]
@sundareswarpullela on `buildkg2c.rtx.ai`, why is the file `/home/ubuntu/RTX/code/kg2c/kg2c_lite.json.gz` only 2.2 MB?  See info that I sent via text and Slack DM.

**Comment by saramsey** [comment]
I have tentative checked off "refreshed DTD" since I heard from @sundareswarpullela who heard from @dkoslicki that we are just going to reuse the KG2.8.0.1c DTD for the purpose of the KG2.8.3c ARAX deployment.

**Comment by sundareswarpullela** [comment]
NGD databases refreshed and the generated `curie_to_pmids.sqlite` and copied to `arax-databases.rtx.ai: /KG2.8.3` and renamed to `curie_to_pmids_v1.0_KG2.8.3.sqlite`

**Comment by chunyuma** [comment]
Hi all,

After discussing with David, we need to rebuild the xDTD model to include "ChemicalEntity" based on the latest version of KG2. It seems like KG2.8.4 hasn't been ready yet. So I might just use KG2.8.3. But no worries, if there is no big change between KG2.8.3 and KG2.8.4, the xDTD database v2.8.3 can be used for KG2.8.4. Any objection?

In addition, I'm wondering if the DTD model (not xDTD) should be retired. We have a discussion before on [issue2052](https://github.com/RTXteam/RTX/issues/2052#issuecomment-1591802749). @dkoslicki, do you have an objection to this point? If we decided to retire it, we may not need the following lines in the checklist for the future KG2:

 refreshed DTD @chunyuma
 DTD model @chunyuma _(may be skipped - depends on the changes in this KG2 version)
 DTD database @chunyuma _(may be skipped - depends on the changes in this KG2 version)

**Comment by saramsey** [comment]
I'm working on this checklist item:

- [ ] turn off the old KG2c version's plover instance

**Comment by saramsey** [comment]
Now working on this checklist item:

- [ ] turn off the new KG2pre version's neo4j instance

Bearing in mind that in this context, "new" means KG2.8.3, and looking at DNS,

<img width="604" alt="Screenshot 2023-07-28 at 11 07 08 AM" src="https://github.com/RTXteam/RTX/assets/5562107/e70ce5a2-cc85-4cc3-97f6-9be0f1aa84fb">

and assuming that "turn off" means shutting down Neo4j and stopping the EC2 instance, it looks like we should be turning off `kg2endpoint3.rtx.ai`.  Checking Neo4j in that endpoint now to make sure it is really KG2.8.3pre:

<img width="518" alt="Screenshot 2023-07-28 at 11 11 41 AM" src="https://github.com/RTXteam/RTX/assets/5562107/0a0f10c0-b7bf-4897-b7d9-c5c2f0e0b6ba">

So, `kg2endpoint3.rtx.ai` is confirmed as the instance that hosts `KG2.8.3pre`.

Now putting in a notification in the `#deployment` channel in the `ARAXTeam` Slack workspace that I will be shutting it down.

OK, with prior notification in the `#deployment` channel, I have shut down Neo4j on `kg2endpoint3.rtx.ai` and stopped the instance:

<img width="628" alt="Screenshot 2023-07-28 at 11 15 54 AM" src="https://github.com/RTXteam/RTX/assets/5562107/82798b52-8ac7-4d2c-9360-784695581780">

**Comment by saramsey** [comment]
Under the aggregate checklist item 

> - [ ] upgrade the ITRB Plover endpoint (https://kg2cploverdb.ci.transltr.io/) to this KG2 version and make the KG2 API start using it (instead of our self-hosted endpoint):

I'm now working on the sub-checklist item:

> - [ ] update kg_config.json in the main branch of the Plover repo to point to the new kg2c_lite_2.X.Y.json.gz file (push this change)

This appears to be already done and pushed to GitHub:

<img width="475" alt="Screenshot 2023-07-28 at 11 23 50 AM" src="https://github.com/RTXteam/RTX/assets/5562107/38f079dc-26ab-40c2-b3bf-d8c836882fff">

**Comment by saramsey** [comment]
I'm now working on the sub-checklist item:

> - [ ]  wait about 45 minutes for the endpoint to rebuild and then run Plover tests to verify it's working

Now, the change to `kg_config.json` was merged from branch `kg2.8.3c` to branch `main` two days ago (i.e., more than 45 minutes ago),

<img width="544" alt="Screenshot 2023-07-28 at 11 30 58 AM" src="https://github.com/RTXteam/RTX/assets/5562107/dbf0a6fe-bd86-43c5-918d-b93277515553">

so in theory we should be good to run the test. However, the test shows that the PloverDB service that is running on `kg2cploverdb.ci.transltr.io` is actually backed by `KG2.8.0c` and not (as I would expect) `KG2.8.3c`:

<img width="1046" alt="Screenshot 2023-07-28 at 11 28 02 AM" src="https://github.com/RTXteam/RTX/assets/5562107/6afd1871-addd-420a-9d62-d24c4c5ed65b">

**Comment by saramsey** [comment]
So, Pouyan says he has not seen any failure of building/deploying PloverDB into CI.

**Comment by edeutsch** [comment]
I wonder if the copying of data from sftp did not happen correctly? Maybe Plover was restarted with new code, but did not undergo a complete redeployment including copying of its database? I'm just speculating wildly, I don't know anything.

**Comment by saramsey** [comment]
Pouyan says his logs indicate that PloverDB deployed successfully into CI on 7/27 (maybe that's GMT), so that tracks with the merge that we did on the morning of 7/26 (PDT).

**Comment by edeutsch** [comment]
do we know that that deployment triggered the download of the 2.8.3 database?

**Comment by saramsey** [comment]
I'm working under the assumption that for whatever reason, it did not download the 2.8.3 database.

Not knowing what else to do, I asked Pouyan to manually trigger a build of PloverDB from `main` and a deployment into ITRB CI.

**Comment by saramsey** [comment]
So, Pouyan reported that the re-deployment of PloverDB into CI was successful. However, I am still seeing:

```
pytest -vs test/test.py::test_version --endpoint https://kg2cploverdb.ci.transltr.io
============================= test session starts ==============================
platform darwin -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0 -- /Users/sramsey/Documents/Work/Proj/PloverDB/venv/bin/python3.11
cachedir: .pytest_cache
rootdir: /Users/sramsey/Documents/Work/Proj/PloverDB
collected 1 item

test/test.py::test_version n00: {'RTX:KG2c': ['RTX-KG2.8.0c', 'biolink:InformationContentEntity', []]}
{'edges': {}, 'nodes': {'n00': {'RTX:KG2c': ['RTX-KG2.8.0c', 'biolink:InformationContentEntity', []]}}}
PASSED

========================================== 1 passed in 3.57s ==========================================
```
```

**Comment by saramsey** [comment]
Whereas on `kg2cplover.rtx.ai`, I am seeing:
```
pytest -vs test/test.py::test_version --endpoint http://kg2cplover.rtx.ai:9990
========================================= test session starts =========================================
platform darwin -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0 -- /Users/sramsey/Documents/Work/Proj/PloverDB/venv/bin/python3.11
cachedir: .pytest_cache
rootdir: /Users/sramsey/Documents/Work/Proj/PloverDB
collected 1 item

test/test.py::test_version n00: {'RTX:KG2c': ['RTX-KG2.8.3c', 'biolink:InformationContentEntity', []]}
{'edges': {}, 'nodes': {'n00': {'RTX:KG2c': ['RTX-KG2.8.3c', 'biolink:InformationContentEntity', []]}}}
PASSED

========================================== 1 passed in 2.52s ==========================================
```

**Comment by saramsey** [comment]
Just now I downloaded the file `https://kg2webhost.rtx.ai/kg2c_lite_2.8.3.json.gz` and checked it, and I can confirm that it is indeed KG2.8.3c:
```
 grep -C 1 RTX:KG2 kg2c_lite_2.8.3.json
    {
      "id": "RTX:KG2",
      "name": "RTX KG2.8.3",
--
    {
      "id": "RTX:KG2c",
      "name": "RTX-KG2.8.3c",
```

**Comment by saramsey** [comment]
Near as I can tell, ITRB PloverDB deploying is not using the latest code in the `main` branch of `RTXteam/PloverDB`.

**Comment by saramsey** [comment]
Still showing as KG2.8.0 in `kg2cploverdb.ci.transltr.io`:
```
pytest -vs test/test.py::test_version --endpoint https://kg2cploverdb.ci.transltr.io
========================================= test session starts =========================================
platform darwin -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0 -- /Users/sramsey/Documents/Work/Proj/PloverDB/venv/bin/python3.11
cachedir: .pytest_cache
rootdir: /Users/sramsey/Documents/Work/Proj/PloverDB
collected 1 item

test/test.py::test_version n00: {'RTX:KG2c': ['RTX-KG2.8.0c', 'biolink:InformationContentEntity', []]}
{'edges': {}, 'nodes': {'n00': {'RTX:KG2c': ['RTX-KG2.8.0c', 'biolink:InformationContentEntity', []]}}}
PASSED

========================================== 1 passed in 2.47s ==========================================
```

**Comment by saramsey** [comment]
> do we know that that deployment triggered the download of the 2.8.3 database?

We don't know that, yet. It's a good idea to check. I have requested a copy of the Jenkins log from ITRB.

**Comment by saramsey** [comment]
So, Pouyan confirms that PloverDB deployment into ITRB CI (i.e., `kg2cploverdb.ci.transltr.io`) uses code in the `main` branch.

**Comment by saramsey** [comment]
On an Ubuntu 18.04 instance with all the local KG2.8.3 databases installed, and with the latest code from the `master` branch (and python3.9.16 compiled from source), I ran the full ARAX test suite (with the default `force_local` set to `False`) and 131 tests passed, 127 skipped, no failures: ```
============================= 136 passed, 127 skipped, 74535 warnings in 595.10s (0:09:55) =============================
```

**Comment by saramsey** [comment]
I believe the old PloverDB endpoint (for KG2.8.0c) has been recycled already to make room for KG2.8.4c (and ARAX is working fine, presumably because `config_dbs.json` is pointing to all ITRB PloverDB services for all three maturity levels), so I'm going to check off this checklist item: 
> turn off our plover endpoint and verify once more that ARAX is still working ok

**Comment by saramsey** [comment]
Now that I have moved the two uncompleted items from this checklist to a separate issue ([RTXteam/RTX issue 2101](https://github.com/RTXteam/RTX/issues/2101)), I am closing this issue out.
